File: quest04/ex05/my_spaceship.c
#include <stdio.h>
#include <string.h>

/*
**
** QWASAR.IO -- my_spaceship
**
** @param {char*} param_1
**
** @return {char*}
**
*/


char* toUpperCase(char* str) {
  int i;
  for (i = 0; i < strlen(str); i++) {
    if (str[i] >= 'a' && str[i] <= 'z') {
      str[i] -= 32; // subtract 32 to convert to uppercase
    }
  }
  return str;
}

char* my_spaceship(char* str) {
  str = toUpperCase(str);
  int i = 0;
  int size = strlen(str);

  int x = 0;
  int y = 0;
  char* direction = "up";

  while (i < size) {
    char mov = str[i];

    if (mov == 'A') {
      if (strcmp(direction, "up") == 0) {
        y--;
      } else if (strcmp(direction, "right") == 0) {
        x++;
      } else if (strcmp(direction, "down") == 0) {
        y++;
      } else if (strcmp(direction, "left") == 0) {
        x--;
      }
    }

    if (mov == 'L') {
      if (strcmp(direction, "up") == 0) {
        direction = "left";
      } else if (strcmp(direction, "left") == 0) {
        direction = "down";
      } else if (strcmp(direction, "down") == 0) {
        direction = "right";
      } else if (strcmp(direction, "right") == 0) {
        direction = "up";
      }
    }

    if (mov == 'R') {
      if (strcmp(direction, "up") == 0) {
        direction = "right";
      } else if (strcmp(direction, "right") == 0) {
        direction = "down";
      } else if (strcmp(direction, "down") == 0) {
        direction = "left";
      } else if (strcmp(direction, "left") == 0) {
        direction = "up";
      }
    }

    i++;
  }

  static char result[100];
  sprintf(result, "{x: %d, y: %d, direction: '%s'}", x, y, direction);
  return result;
}

File: my_spaceship/ex00/my_spaceship.js
function my_spaceship(str) {
    // str = str.toUpperCase();
    let i = 0;
    let size = str.length;
  
    let x = 0;
    let y = 0;
    let direction = "up";
  
    while (i < size) {
      mov = str[i];
  
      if (mov === "A") {
        if (direction === 'up') (y--);
        else if (direction === 'right') (x++);
        else if (direction === 'down') (y++);
        else if (direction === 'left') (x--);
      }
  
      if (mov === "L") {
        if (direction === 'up') ((direction) = 'left');
        else if (direction === 'left') ((direction) = 'down');
        else if (direction === 'down') ((direction) = 'right');
        else if (direction === 'right') ((direction) = 'up');
      }
  
      if (mov === "R") {
        if (direction === 'up') ((direction) = 'right');
        else if (direction === 'right') ((direction) = 'down');
        else if (direction === 'down') ((direction) = 'left');
        else if (direction === 'left') ((direction) = 'up');
      }
      i++;
    }
    return "{x: " + x + ", y: " + y + ", direction: '" + direction + "'}";
  }